Sources/NIOSSH/Keys And Signatures/NIOSSHPublicKey.swift

```diff
@@ -175,16 +187,100 @@ extension NIOSSHPublicKey {
     internal static var knownAlgorithms: [String.UTF8View] {
-        [Self.ed25519PublicKeyPrefix, Self.ecdsaP384PublicKeyPrefix, Self.ecdsaP256PublicKeyPrefix, Self.ecdsaP521PublicKeyPrefix]
+        bundledAlgorithms + customPublicKeyAlgorithms.map { $0.publicKeyPrefix.utf8 }
+    }
+    
+    internal static var customPublicKeyAlgorithms: [NIOSSHPublicKeyProtocol.Type] {
+        customAlgorithmsLock.lock()
+        defer { customAlgorithmsLock.unlock() }
+        return _customPublicKeyAlgorithms
+    }
+    
+    internal static var customSignatures: [NIOSSHSignatureProtocol.Type] {
+        customAlgorithmsLock.lock()
+        defer { customAlgorithmsLock.unlock() }
+        return _customSignatures
     }
 }
 
+public enum NIOSSHAlgorithms {
+    public static func register(keyExchangeAlgorithm type: NIOSSHKeyExchangeAlgorithmProtocol.Type) {
+        customAlgorithmsLock.lock()
+        defer { customAlgorithmsLock.unlock() }
+        
+        if !_customKeyExchangeAlgorithms.contains(where: { ObjectIdentifier($0) == ObjectIdentifier(type) }) {
+            _customKeyExchangeAlgorithms.append(type)
+        }
+    }
+    
+    public static func register(transportProtectionScheme type: NIOSSHTransportProtection.Type) {
+        customAlgorithmsLock.lock()
+        defer { customAlgorithmsLock.unlock() }
+        
+        if !_customTransportProtectionSchemes.contains(where: { ObjectIdentifier($0) == ObjectIdentifier(type) }) {
+            _customTransportProtectionSchemes.append(type)
+        }
+    }
+    
+    /// Registers a custom type tuple for use in Public Key Authentication.
+    public static func register<
+        PublicKey: NIOSSHPublicKeyProtocol,
+        Signature: NIOSSHSignatureProtocol
+    >(
+        publicKey type: PublicKey.Type,
+        signature: Signature.Type
+    ) {
+        customAlgorithmsLock.lock()
+        defer { customAlgorithmsLock.unlock() }
+        
+        if !_customPublicKeyAlgorithms.contains(where: { ObjectIdentifier($0) == ObjectIdentifier(type) }) {
+            _customPublicKeyAlgorithms.append(type)
+            _customSignatures.append(signature)
+        }
+    }
+    
+    /// Used for our unit tests
+    internal static func unregisterAlgorithms() {
+        customAlgorithmsLock.lock()
+        defer { customAlgorithmsLock.unlock() }
+        
+        _customKeyExchangeAlgorithms = []
+        _customSignatures = []
+        _customPublicKeyAlgorithms = []
+        _customTransportProtectionSchemes = []
+    }
+}
+
+internal var customTransportProtectionSchemes: [NIOSSHTransportProtection.Type] {
+    customAlgorithmsLock.lock()
+    defer { customAlgorithmsLock.unlock() }
+    return _customTransportProtectionSchemes
+}
+
+internal var customKeyExchangeAlgorithms: [NIOSSHKeyExchangeAlgorithmProtocol.Type] {
+    customAlgorithmsLock.lock()
+    defer { customAlgorithmsLock.unlock() }
+    return _customKeyExchangeAlgorithms
+}
+
+internal let customAlgorithmsLock = NSLock()
```

Comment:
We should probably use fine-grained locking here, so can we have one lock per field? It'll reduce contention on the locks.

Reply:
Additionally, instead of `NSLock` can we add a dependency on `NIOConcurrencyHelpers` and use that `Lock`?